R/class-plotdata-box.R

```diff
@@ -183,45 +200,73 @@ box.dt <- function(data, map, points = c('outliers', 'all', 'none'), mean = c(FA
     data.table::setDT(data)
   }
 
-  # Handle repeated plot references
+  map <- validateMap(map)
+
+  # If there is a duplicated plotRef in map, it must match listVarPlotRef
   if (any(duplicated(map$plotRef))) {
-    
-    # Identify the list var based on any plotRef that is repeated
-    listVarPlotRef <- unique(map$plotRef[duplicated(map$plotRef)])
-    listVarPlotRef <- validateListVar(map, listVarPlotRef)
-    
-    # Box-specific
-    if (listVarPlotRef == 'xAxisVariable' | listVarPlotRef == 'facetVariable1') {
-      meltedValuePlotRef <- 'yAxisVariable'
-    } else {
-      stop("Incompatable repeated variable")
-    }
-    
-    # Check to ensure meltedValuePlotRef is not already defined
-    if (any(map$plotRef == meltedValuePlotRef)) {
-      stop(paste0("Cannot melt data: ", meltedValuePlotRef, " already defined."))
+    if (!identical(listVarPlotRef, unique(map$plotRef[duplicated(map$plotRef)]))) {
+      stop('listVar error: duplicated map plotRef does not match listVarPlotRef.')
     }
+  }
 
-    # Record variable order
-    listVarIdOrder <- map$id[map$plotRef == listVarPlotRef]
-    
-    # Melt data and update the map 
-    data <- data.table::melt(data, measure.vars = listVarIdOrder, variable.factor = FALSE, variable.name='meltedVariable', value.name='meltedValue')
-    map <- remapListVar(map, listVarPlotRef, meltedValuePlotRef)
-    
-  } # end handling of repeated plot element references
+  # If listVar and inferredVar labels are provided, must also provide listVarPlotRef
+  if ((!is.null(listVarDisplayLabel) | !is.null(inferredVarDisplayLabel)) & is.null(listVarPlotRef)) {
+    stop('listVar error: listVarPlotRef must be specified in order to use inferredVarDisplayLabel or listVarDisplayLabel')
+  }
 
   xAxisVariable <- plotRefMapToList(map, 'xAxisVariable')
   if (is.null(xAxisVariable$variableId)) {
     stop("Must provide xAxisVariable for plot type box.")
   }
   yAxisVariable <- plotRefMapToList(map, 'yAxisVariable')
-  if (is.null(yAxisVariable$variableId)) {
+  if (is.null(yAxisVariable$variableId) & is.null(listVarPlotRef)) {
     stop("Must provide yAxisVariable for plot type box.")
   }
   overlayVariable <- plotRefMapToList(map, 'overlayVariable')
   facetVariable1 <- plotRefMapToList(map, 'facetVariable1')
   facetVariable2 <- plotRefMapToList(map, 'facetVariable2')
+  
+  # Handle listVars
+  listVarDetails <- list('inferredVariable' = NULL,
+                         'inferredVarPlotRef' = 'yAxisVariable',
+                         'listVarPlotRef' = listVarPlotRef,
+                         'listVarDisplayLabel' = listVarDisplayLabel)
+
+  if (!is.null(listVarPlotRef)) {
+    if (identical(listVarPlotRef, 'xAxisVariable')) {
+      
+      # Ensure all variables are numbers
+      if (!all(xAxisVariable$dataType == 'NUMBER')){
+        stop("listVar error: All overlay vars must be of type NUMBER.")
+      }
```

Comment:
in the current scoped uses for listvars, yes. If we let them work with other kinds of data, then I think it would need to be plot-specific. For example, if we let a user choose a listvar for the bar x axis, then we'd need to ensure the dataShape was categorical.
I'll move to plot.data class until we hit that scenario, though. It'll be easy to split up if we come to that. I'll aslo fix that error message :)